primitives/state-machine/src/overlayed_changes/offchain.rs

```diff
@@ -0,0 +1,254 @@
+		OffchainOverlayedChangesIntoIter::drain(self)
+	}
+
+	/// Remove a key and its associated value from the offchain database.
+	pub fn remove(&mut self, prefix: &[u8], key: &[u8]) {
+		if let Self::Enabled(storage) = self {
+			let _ = storage.set(
+				(prefix.to_vec(), key.to_vec()),
+				OffchainOverlayedChange::Remove,
+				None,
+			);
+		}
+	}
+
+	/// Set the value associated with a key under a prefix to the value provided.
+	pub fn set(&mut self, prefix: &[u8], key: &[u8], value: &[u8]) {
+		if let Self::Enabled(storage) = self {
+			let _ = storage.set(
+				(prefix.to_vec(), key.to_vec()),
+				OffchainOverlayedChange::SetValue(value.to_vec()),
+				None,
+			);
+		}
+	}
+
+	/// Obtain a associated value to the given key in storage with prefix.
+	pub fn get(&self, prefix: &[u8], key: &[u8]) -> Option<OffchainOverlayedChange> {
+		if let Self::Enabled(storage) = self {
+			let key = (prefix.to_vec(), key.to_vec());
+			storage.get(&key).map(|entry| entry.value_ref()).cloned()
+		} else {
+			None
+		}
+	}
+
+	/// Reference to inner change set.
+	pub fn overlay(&self) -> Option<&OverlayedMap<(Vec<u8>, Vec<u8>), OffchainOverlayedChange>> {
+		if let Self::Enabled(storage) = self {
+			Some(&storage)
+		} else {
+			None
+		}
+	}
+
+	/// Mutable reference to inner change set.
+	pub fn overlay_mut(&mut self) -> Option<&mut OverlayedMap<(Vec<u8>, Vec<u8>), OffchainOverlayedChange>> {
+		if let Self::Enabled(storage) = self {
+			Some(storage)
+		} else {
+			None
+		}
+	}
+
+}
+
+use sp_std::collections::btree_map;
+
+/// Iterate by reference over the prepared offchain worker storage changes.
+pub struct OffchainOverlayedChangesIter<'i> {
+	inner: Option<sp_std::iter::Map<
+		btree_map::Iter<'i, (Vec<u8>,Vec<u8>), OverlayedEntry<OffchainOverlayedChange>>,
+		fn((&'i (Vec<u8>, Vec<u8>), &'i OverlayedEntry<OffchainOverlayedChange>))
+			-> (&'i (Vec<u8>, Vec<u8>), &'i OffchainOverlayedChange),
+	>>,
+
+}
+
+impl<'i> Iterator for OffchainOverlayedChangesIter<'i> {
+	type Item = (&'i (Vec<u8>, Vec<u8>), &'i OffchainOverlayedChange);
+	fn next(&mut self) -> Option<Self::Item> {
+		if let Some(ref mut iter) = self.inner {
+			iter.next()
+		} else {
+			None
+		}
+	}
+}
+
+impl<'i> OffchainOverlayedChangesIter<'i> {
+	/// Create a new iterator based on a refernce to the parent container.
+	pub fn new(container: &'i OffchainOverlayedChanges) -> Self {
+		match container {
+			OffchainOverlayedChanges::Enabled(inner) => Self {
+				inner: Some(inner.changes.iter().map(map_kv_ref))
+			},
+			OffchainOverlayedChanges::Disabled => Self { inner: None, },
+		}
+	}
+}
+
+
+/// Iterate by value over the prepared offchain worker storage changes.
+pub struct OffchainOverlayedChangesIntoIter {
+	inner: Option<sp_std::iter::Map<
+		btree_map::IntoIter<(Vec<u8>,Vec<u8>), OverlayedEntry<OffchainOverlayedChange>>,
+		fn(((Vec<u8>, Vec<u8>), OverlayedEntry<OffchainOverlayedChange>))
+			-> ((Vec<u8>, Vec<u8>), OffchainOverlayedChange),
```

Comment:
Again, deserves a bunch of typedefs. Especially the `(Vec<u8>, Vec<u8>)` stuff across the entire file.